### PR TITLE
feat(codeowners): Async update Code Owners on association updates

### PR DESCRIPTION
**Pull request by NisanthanNanthakumar**
Objective:
Now, we will allow users to upload CodeOwners with incomplete external mappings. So when they start to fill out the missing mappings, we should automatically update our `sentry_projectcodeowners.schema` to use the new mappings. 

We will trigger an async task to re-evaluate the most recent CODEOWNERS raw with the new mappings using Django signals `post_save` and `post_delete` for ExternalActor, Team and ProjectTeam models.

### CODE REVIEW

File: src/sentry/models/project.py [hunk 1]
```diff
@@ -435,3 +436,17 @@ def delete(self, **kwargs):
 
 
 pre_delete.connect(delete_pending_deletion_option, sender=Project, weak=False)
+post_save.connect(
+    lambda instance, **kwargs: update_code_owners_schema.apply_async(
+        kwargs={"organization": None, "integration": None, "projects": [instance.project]}
```
Comment (MeredithAnya) [inline src/sentry/models/project.py:441]: we can get the `organization` from the `instance.project.organization`. I think we should plan on always passing in a organization to the task

File: src/sentry/tasks/code_owners.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+logger = logging.getLogger("sentry.tasks.code_owners")
+
+
+@instrumented_task(
+    name="sentry.tasks.update_code_owners_schema",
+    queue="code_owners",
+    default_retry_delay=5,
+    max_retries=5,
+)
+def update_code_owners_schema(organization=None, integration=None, projects=None, **kwargs):
```
Comment (MeredithAnya) [inline src/sentry/tasks/code_owners.py:15]: we can change the arguments to always expect an organization (and not have the default be `None`)

File: src/sentry/tasks/code_owners.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    if not features.has("organizations:integrations-codeowners", organization):
+        return
+    try:
+        code_owners = []
+
+        if projects:
+            code_owners = ProjectCodeOwners.objects.filter(project__in=projects)
+
+        if organization and integration:
```
Comment (MeredithAnya) [inline src/sentry/tasks/code_owners.py:26]: then here could just be checking for the integration. basically it looks like either you have the `projects` or you have the `integration`

File: src/sentry/models/team.py [hunk 4]
```diff
@@ -250,3 +270,12 @@ def get_projects(self):
         from sentry.models import Project
 
         return Project.objects.get_for_team_ids({self.id})
+
+    def delete(self, **kwargs):
+        from sentry.models import Actor, ExternalActor
+
+        # There is no foreign key relationship so we have to manually cascade.
+        ExternalActor.objects.filter(actor=self.actor).delete()
+        Actor.objects.filter(id=self.actor_id).delete()
```
Comment (MeredithAnya) [inline src/sentry/models/team.py:279]: The `ExternalActor` should get deleted when the `Actor` is deleted right? So do we need both of these here? can we just keep `Actor.objects.filter(id=self.actor_id).delete()`?

File: src/sentry/tasks/code_owners.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import logging
+
+from sentry import features
+from sentry.tasks.base import instrumented_task
+
+logger = logging.getLogger("sentry.tasks.code_owners")
```
Comment (MeredithAnya) [inline src/sentry/tasks/code_owners.py:6]: nit: it seems like we use `codeowners` as one word in many places without the `_`, do we want to be consistent with for this file as well?

Comment (NisanthanNanthakumar) [reply src/sentry/tasks/code_owners.py:6]: @MeredithAnya We're calling the feature `Code Owners`. So I'm migrating to using `codeowners` to denote the CODEOWNERS text and `code_owners` for our feature.

File: src/sentry/models/team.py [hunk 6]
```diff
@@ -85,6 +86,23 @@ def get_for_user(self, organization, user, scope=None, with_projects=False):
 
         return results
 
+    def post_save(self, instance, **kwargs):
+
```
Comment (MeredithAnya) [inline src/sentry/models/team.py:90]: nit: can we get rid of this empty line just to make it consistent with below

File: src/sentry/tasks/code_owners.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            ).values_list("id", flat=True)
+
+            code_owners = ProjectCodeOwners.objects.filter(
+                repository_project_path_config__in=code_mapping_ids
+            )
+
+        for code_owner in code_owners:
+            code_owner.update_schema()
+
+    except (RepositoryProjectPathConfig.DoesNotExist, ProjectCodeOwners.DoesNotExist):
```
Comment (MeredithAnya) [inline src/sentry/tasks/code_owners.py:39]: can we add a todo here or somewhere else to remind us to come back and logging for the cases where we might want to have more information if something fails?

File: src/sentry/tasks/code_owners.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+logger = logging.getLogger("sentry.tasks.code_owners")
+
+
+@instrumented_task(
+    name="sentry.tasks.update_code_owners_schema",
+    queue="code_owners",
+    default_retry_delay=5,
+    max_retries=5,
+)
+def update_code_owners_schema(organization, integration=None, projects=None, **kwargs):
```
Comment (MeredithAnya) [inline src/sentry/tasks/code_owners.py:15]: Also can we add some tests for this task?

**Review by MeredithAnya — APPROVED** [review]
once tests are passing (including new ones for the task) should be good to go. going to approve it now to unblock

**Pull request merged**